### ISSUE TITLE
'Heavy' Size PDFs crashing

### ISSUE THREAD

**Issue by davidvelosa** [koreader/koreader#7627]
[crash.log](https://github.com/koreader/koreader/files/6411897/crash.log)
* KOReader version: 2021.3
* Device: Kobo H2O

#### Issue
I open this PDF file (about 23-30MB), it's a manga so it's based on images, the file is about 22- 24 pages, so when I'm on page 18-20 KoReader crashes (shows up a log with a bomb) and restarts again from the previous file I had open. So I have to look again for the file and fresh open it, the position I was on doesn't get saved. This keeps happening to me with different files of about the same weight, I attach the crash.log if it serves for anything. And should you need something else, just let me know. Thanks for the good work Koreader Team :)

PD: The same thing happened to me when I tried to update Koreader while having the PDF opened

#### Steps to reproduce
Open the PDF and just keep reading through all the pages until it crashes

##### `crash.log` (if applicable)

I put some samples below, but the full file is here if you want to see it:

<details>
  <summary>crash.log</summary>

```
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 04/29/21-22:04:21
 [*] Version: v2021.03

ffi.load: blitbuffer
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): libSDL2-2.0.so.0: cannot open shared object file: No such file or directory
SDL2 not loaded:	./setupkoenv.lua:27: Not able to load dynamic library: libSDL2-2.0.so.0
04/29/21-22:04:21 INFO  initializing for device Kobo_storm
04/29/21-22:04:21 INFO  framebuffer resolution: {
    ["w"] = 1264,
    ["h"] = 1680,
}
[ko-input] Forked off fake event generator(pid:15878).
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
04/29/21-22:04:22 INFO  opening file /mnt/onboard/manga/GTO/Chapter 16.pdf
ffi.load: libs/liblept.so.5
ffi.load: libs/libk2pdfopt.so.2
04/29/21-22:04:23 INFO  Loading plugins from directory: plugins
04/29/21-22:04:23 INFO  Plugin  plugins/autofrontlight.koplugin/main.lua  has been disabled.
04/29/21-22:04:23 INFO  Plugin  plugins/autostandby.koplugin/main.lua  has been disabled.
ffi.load: sqlite3
ffi.load: libs/libzstd.so.1
04/29/21-22:04:23 INFO  Plugin  plugins/coverimage.koplugin/main.lua  has been disabled.
04/29/21-22:04:23 INFO  Plugin  plugins/hello.koplugin/main.lua  has been disabled.
04/29/21-22:04:23 INFO  RD loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin opds at plugins/opds.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:04:23 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:04:23 INFO  setting zoom mode to page
04/29/21-22:04:27 INFO  event times are real clock time: no adjustment needed
04/29/21-22:05:05 INFO  setting zoom mode to manual
04/29/21-22:05:32 INFO  setting zoom mode to page
04/29/21-22:05:44 INFO  FM loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin opds at plugins/opds.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:05:44 INFO  FM loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:05:46 INFO  opening file /mnt/onboard/manga/GTO/Chapter 17.pdf
04/29/21-22:05:47 INFO  RD loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin opds at plugins/opds.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:05:47 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:05:47 INFO  setting zoom mode to page
04/29/21-22:05:47 INFO  setting zoom mode to pagewidth
04/29/21-22:07:06 INFO  setting zoom mode to page
04/29/21-22:13:14 INFO  WakeupMgr: scheduling wakeup in 259200
04/29/21-22:13:15 INFO  Kobo resume: clean up after wakeup
04/29/21-22:13:15 INFO  Kobo resume: unflagged kernel subsystems for resume, ret: true
04/29/21-22:14:04 INFO  setting zoom mode to page
Killed
!!!!
Uh oh, something went awry... (Crash n°1: 04/29/21 @ 22:15:49)
Running FW 4.26.16704 on Linux 4.1.15-00416-gfc725f43860b (#99 SMP PREEMPT Mon Apr 6 14:55:22 CST 2020)
Attempting to restart KOReader . . .
!!!!
---------------------------------------------
                launching...

===

     launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 04/29/21-22:15:51
 [*] Version: v2021.03

ffi.load: blitbuffer
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): libSDL2-2.0.so.0: cannot open shared object file: No such file or directory
SDL2 not loaded:	./setupkoenv.lua:27: Not able to load dynamic library: libSDL2-2.0.so.0
04/29/21-22:15:51 INFO  initializing for device Kobo_storm
04/29/21-22:15:51 INFO  framebuffer resolution: {
    ["w"] = 1264,
    ["h"] = 1680,
}
[ko-input] Forked off fake event generator(pid:15916).
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
04/29/21-22:15:52 INFO  opening file /mnt/onboard/manga/GTO/Chapter 17.pdf
ffi.load: libs/liblept.so.5
ffi.load: libs/libk2pdfopt.so.2
04/29/21-22:15:52 INFO  Loading plugins from directory: plugins
04/29/21-22:15:52 INFO  Plugin  plugins/autofrontlight.koplugin/main.lua  has been disabled.
04/29/21-22:15:52 INFO  Plugin  plugins/autostandby.koplugin/main.lua  has been disabled.
ffi.load: sqlite3
ffi.load: libs/libzstd.so.1
04/29/21-22:15:52 INFO  Plugin  plugins/coverimage.koplugin/main.lua  has been disabled.
04/29/21-22:15:52 INFO  Plugin  plugins/hello.koplugin/main.lua  has been disabled.
04/29/21-22:15:53 INFO  RD loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin opds at plugins/opds.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:15:53 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:15:53 INFO  setting zoom mode to manual
04/29/21-22:15:59 INFO  event times are real clock time: no adjustment needed
04/29/21-22:15:59 INFO  setting zoom mode to page
04/29/21-22:17:17 INFO  FM loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin opds at plugins/opds.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:17:17 INFO  FM loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:17:21 INFO  opening file /mnt/onboard/manga/GTO/Chapter 18.pdf
04/29/21-22:17:22 INFO  RD loaded plugin autosuspend at plugins/autosuspend.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin backgroundrunner at plugins/backgroundrunner.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin batterystat at plugins/batterystat.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin calibre at plugins/calibre.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin coverbrowser at plugins/coverbrowser.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin docsettingtweak at plugins/docsettingtweak.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin gestures at plugins/gestures.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin keepalive at plugins/keepalive.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin kosync at plugins/kosync.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin movetoarchive at plugins/movetoarchive.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin opds at plugins/opds.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin perceptionexpander at plugins/perceptionexpander.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin profiles at plugins/profiles.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin qrclipboard at plugins/qrclipboard.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin readtimer at plugins/readtimer.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin statistics at plugins/statistics.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin systemstat at plugins/systemstat.koplugin
04/29/21-22:17:22 INFO  RD loaded plugin timesync at plugins/timesync.koplugin
04/29/21-22:17:22 INFO  setting zoom mode to page
04/29/21-22:17:22 INFO  setting zoom mode to pagewidth
04/29/21-22:17:24 INFO  setting zoom mode to page
04/29/21-22:17:37 INFO  WakeupMgr: scheduling wakeup in 259200
04/29/21-22:17:39 INFO  Kobo resume: clean up after wakeup
04/29/21-22:17:39 INFO  Kobo resume: unflagged kernel subsystems for resume, ret: true
04/29/21-22:18:02 INFO  setting zoom mode to page
04/29/21-22:18:44 INFO  setting zoom mode to manual
04/29/21-22:19:01 INFO  setting zoom mode to page
Killed
!!!!
Uh oh, something went awry... (Crash n°1: 04/29/21 @ 22:23:31)
Running FW 4.26.16704 on Linux 4.1.15-00416-gfc725f43860b (#99 SMP PREEMPT Mon Apr 6 14:55:22 CST 2020)
Attempting to restart KOReader . . .
!!!!
```
</details>

**Comment by NiLuJe** [comment]
OOM kill, not much we can do about it. Although is *is* slightly unexpected for a relatively small file on a device with 512MB of RAM to ourselves.

Any specific settings that might explain this? (e.g., reflow, weird zoom modes or something)?

You can check the current RAM usage with the corresponding widget in the Status bar, FWIW.

**Comment by Frenzie** [comment]
Is the file something that can be shared? Which settings are you using with it?

**Comment by NiLuJe** [comment]
Yeah, if not, we could certainly use a look at it, you can send it to us privately (either via Gitter or MobileRead).

You mentioned manga, is it a CBZ, or actually a PDF? And, if so, is it one of those high-res PDF that requires seven billion layers to compose a single panel? Because that could explain the ballooning RAM usage.

**Comment by davidvelosa** [comment]
So about settings, I have it preset to Fit: Full, but when I open a any PDF it starts with Fit: Width, so I have to manually change to full (zoom out), then in normal use when the text is too small I do Zoom: percentage - zoom factor: 2, and eventually I put it back to full, this I don't do it in all pages though, let's say three of them . I don't use reflow

I just turned on Memory Usage in the Status bar and I'm noticing big jumps in the memory:
(starts with) 28 => 30 => (here I open a fresh file) 50  => (from here on I start moving through pages normally, without doing zoom or anything) 88 => 106 => 124 => 142 => 154 => 172 => 191 => 209 => 227 => 246 => 264 => 282 => 301 => 319 => 337 => 356 => 375 => 393 => 411 =>  ? (at page 20 it crashes)
I'm only moving forward through pages, so at this time I don't go backwards or anything

It's a PDF file not CBZ, I haven't checked about the layers which could make sense but I can't check at this moment. I will send you the file, please let me know how can I do it either through Gitter or Mobileread. Would Wetransfer work?

**Comment by Frenzie** [comment]
Wetransfer should be fine, maybe better to send it to @NiLuJe since I'll be pretty busy driving around the country this week. Or just to use both.

> 88 => 106 => 124 => 142 => 154 => 172 => 191 => 209 => 227 => 246 => 264 => 282 => 301 => 319 => 337 => 356 => 375 => 393 => 411 => ? (at page 20 it crashes)

The first few of those are (probably) normal, since there's a cache (so you can quickly go back to a limited number of previous pages). But that behavior is certainly very odd and might suggest a different kind of memory leak.

https://github.com/koreader/koreader/blob/0d6be41c75227aacdfff83f5c50cf5eb5a6f9032/defaults.lua#L29-L36

**Comment by davidvelosa** [comment]
Link to file:
https://we.tl/t-3LkxIRIaKS

**Comment by davidvelosa** [comment]
> https://github.com/koreader/koreader/blob/0d6be41c75227aacdfff83f5c50cf5eb5a6f9032/defaults.lua#L29-L36


I just checked on this file and the lines are the same on mine

**Comment by Frenzie** [comment]
The "good" news is I can confirm the issue. @NiLuJe I wonder if there might be a regression somewhere? I haven't read any documents like this recently but in the past I've read them without any issues.

**Comment by NiLuJe** [comment]
In the rare few cases I do that, I usually rip the PDF/ePub apart and build a target-specific CBZ out of it, but, yeah, something's definitely going wrong somewhere here ;).

**Comment by NiLuJe** [comment]
It vaguely seems like TileCacheItems are pinning the source objects in memory too, somehow, which doesn't make much sense at a glance...

But if I stick a random * 16 to Document's tile sizes, I do ultimately get a massive gc sweep at ~400MB occupied.

**Comment by NiLuJe** [comment]
Hmm, unless the pgdim/pglinks objects are actual refs to FFI objects, and not copies...

**Comment by NiLuJe** [comment]
That theory seems to hold some water.

Neutering `PdfDocument:getPageLinks`, which, at a very quick glance, seems to store refs to FFI objects, yields to a saner behavior. (e.g., I get a GC sweep at ~300MB occupied, which is still far more than the 60MB of Cache)

**Comment by davidvelosa** [comment]
From my _non-connoisseur_ point of view I'm guessing:

1. There something telling the device that the file can take an unlimited file size of cache
2. The device is caching the same information every new page (taking in account it might cache previous and next pages to display faster) but at the same time is assuming is different from the information cached in the previous page. I don't know if that makes sense to you LOL. What makes me think this, is that memory is being taken in 18mb chunks from a 35mb file every page turn, which seems like a loop. And this probably you won't ever notice with lighter and shorter length files

So maybe adding something that purges cache in large files every 5-6 pages can do the job, or rewriting whatever makes 2. happen, but I'm assuming is from cache as well

**Comment by NiLuJe** [comment]
On a H2O, a single screen is worth ~1.5MB of data. That's, on paper, the single largest element we'll *explicitly* cache.

But, of course, this is more complex than that, and we also cache bits of metadata from the rendering engine itself, and apparently some of it forces the engine's own data to be pinned to memory for far longer than expected.

I don't really see a way to estimate that sanely, so, best I can figure out would be to actually check the live RAM watermark on cache insertion, instead of blindly trusting *only* the cache's watermark.

**Comment by davidvelosa** [comment]
I see, makes sense

**Comment by NiLuJe** [comment]
Well, it's definitely somewhere deep in MµPDF, and somehow playing with the Cache may nudge it to do the right thing, somehow.

Random observations:

* Switching to our custom debugging allocators (debug_memory = true in ffi/mupdf.lua) yields somewhat better results, but confirms that MuPDF is happily munching on memory well past the 8MB the context should limit itself to.
* Attempting to free the context "early" (i.e., on document close) segfaults. Haven't looked at a debug build to see why.
* Closing the Document doesn't release the memory, but opening a *different* *PDF* document does. (That's what my "kill the context with fire" experiment wanted to double-check).

**Comment by NiLuJe** [comment]
Might be fixed in modern MuPDF versions. `>_<"`.

**Comment by davidvelosa** [comment]
Another thing that I've noted is when you skim between pages, it doesn't do the same effect on memory, and you can move 'more freely' through pages since it doesn't take the same amount of memory chunks. As well when zooming the page, makes variations on the chunks of memory (zoom in and out, and go forward or backwards).

**Comment by NiLuJe** [comment]
Yep, if stuff's in the cache, we don't hit MµPDF *at all* (almost).

So, on a clean run, I can go through the full PDF, then open *another* PDF for a few pages to reclaim memory, then open the manga, and I'll happily sit at ~100MB with almost zero fluctuations because it's 100% of cache hits.

**Comment by NiLuJe** [comment]
Yeah, definitely an old MµPDF bug related to its store cache. Doubling the max store cache magically hides the issue for now...

**Comment by NiLuJe** [comment]
As a quick "more testers always good", changing [this line](https://github.com/koreader/koreader-base/blob/85dacfbba80e4f7a540ef9de3dff1186440d707b/ffi/mupdf.lua#L31) to 16MB instead was the single most effective thing at sweeping this under the rug.

I mean, it probably still leaks/corrupts memory (as dropping the context still implodes), but it's way less noticeable.

**Comment by Frenzie** [comment]
> In the rare few cases I do that, I usually rip the PDF/ePub apart

Then you lose the vectors and you get a clearly inferior, blurrier experience (unless there are none, of course).

**Comment by poire-z** [comment]
> possibly related to the weird leaky feelings you've expressed about Scroll*Widgets that use MµPDF, because, in technical terms, it looks like shit is hella broke :D 

I think this was about re-using an existing MuPdf wrapper object and setContent(next_html_dict_content) when switching between HTML dict results ?
Dunno, I had seen the possbile memory leaks before you changed this to doing that - and it's the HTML rendering part of MuPDF that is at play there, so possible not exposing the same kind of issue you're hunting here.

No other thought about this issue though :/ 

> Might be fixed in modern MµPDF versions.

Better finish https://github.com/koreader/koreader-base/pull/1203 to find out - rather than spending much time on our old version :)

**Comment by Frenzie** [comment]
Agreed on that one.

**Comment by NiLuJe** [comment]
true enough, but, do note that, Ironically enough, not modern enough ;p (I'm not even quite sure it's entirely fixed in 1.17, I haven't checked the release date compared to master).

**Comment by davidvelosa** [comment]
> As a quick "more testers always good", changing [this line](https://github.com/koreader/koreader-base/blob/85dacfbba80e4f7a540ef9de3dff1186440d707b/ffi/mupdf.lua#L31) to 16MB instead of 8MB was the single most effective attempt at sweeping this under the rug of all the night's experiments.
> 

@NiLuJe Just did this change, and it seems to work better and leak much less memory, it's even reducing the memory take in some points. At least is not crashing anymore, so I will do the test and let you guys know if I get further problems. Thanks for your help guys :)

**Comment by davidvelosa** [comment]
Just to let you know, this seems to solve another two issues that I previously had with the system crashing when these kinds PDFs are open:
- When trying to update (this one is random)
- When trying to modify bookmark notes in PDFs (in this one would always crash, no matter the pdf)

**Comment by NiLuJe** [comment]
Fun fact: KOptContext caching was broken and nobody noticed. Yay, PDF.

**Comment by NiLuJe** [comment]
"Fixed" via #7624 ;).
